### PR TITLE
Task2

### PR DESCRIPTION
**Pull request by Kvazar103**
Task2

### CODE REVIEW

File: force-app/main/default/triggers/CarTrigger.trigger [hunk 1]
```diff
@@ -0,0 +1,24 @@
+trigger CarTrigger on Product2 (before insert,after update) {
+    
+    Trigger_Control__c triggerSetting=Trigger_Control__c.getInstance('MakeActive');
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/CarTrigger.trigger:3]: **NO CHANGE NEEDED, JUST SUGGESTION**

It makes sense to name custom setting records with more meaningful names, like CarTriggerSettings or smth like that.

It's better to have the checkbox isDisabled rather then isActive. If you remember, custom settings are not deployable, so each time you will have to manually create them and set checkbox for ALL triggers you want to have active, and usually it's ALL triggers in the org. So, it's just easier to select which one you want to disable, rather then enable all of them and then diselect

File: force-app/main/default/triggers/CarTrigger.trigger [hunk 2]
```diff
@@ -0,0 +1,24 @@
+trigger CarTrigger on Product2 (before insert,after update) {
+    
+    Trigger_Control__c triggerSetting=Trigger_Control__c.getInstance('MakeActive');  
+    
+    if((triggerSetting!=null) && (triggerSetting.isActive__c)){
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/CarTrigger.trigger:5]: **_NO CHANGE NEEDED, JUST SUGGESTION_**

Following the same logic, if you do isDisabled, you want to run trigger even is trigger settings is null, this condition `(triggerSetting!=null) && (triggerSetting.isActive__c)` will be changed to `(triggerSetting == null) || triggerSetting.isDisabled__c`

File: force-app/main/default/triggers/CarTrigger.trigger [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        try{
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/CarTrigger.trigger:6]: Do not add error handling inside trigger itself, move it to handler method if it's needed

File: force-app/main/default/triggers/OpportunityLineItemTrigger.trigger [hunk 4]
```diff
@@ -0,0 +1,24 @@
+trigger OpportunityLineItemTrigger on OpportunityLineItem (before delete,after insert,after update) {
+    
+   Trigger_Control__c triggerSetting=Trigger_Control__c.getInstance('MakeActive');  
+    
+    if((triggerSetting!=null) && (triggerSetting.isActive__c)){
+        try{
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/OpportunityLineItemTrigger.trigger:6]: Same here, please move error handling to handler methods

File: force-app/main/default/triggers/CarTrigger.trigger [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+            if(Schema.sObjectType.Product2.isAccessible()){//if user have access to object and fields
+                if(Trigger.isAfter && Trigger.isUpdate){
+                    CarTriggerHelper.updateCar(Trigger.oldMap);
+                }  
+            }else{
+                system.debug('The user doesnt have access to this object');
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/CarTrigger.trigger:12]: You addError here to notify user why trigger didn't work, because for now it's just not executed, end for customer is not really clear why

File: force-app/main/default/triggers/OpportunityLineItemTrigger.trigger [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+            if(Schema.sObjectType.OpportunityLineItem.isAccessible()){//if user have access to object and fields
+                if(Trigger.isAfter && Trigger.isInsert){
+                    OpportunityLineItemTriggerHelper.checkCarStatusBeforeInsert();
+                }else if(Trigger.isDelete && Trigger.isBefore){
+                    OpportunityLineItemTriggerHelper.deleteOpportunityLineItem();
+                }
+            }else{
+                system.debug('The user doesnt have access to this object');
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/OpportunityLineItemTrigger.trigger:14]: You addError here to notify user why trigger didn't work, because for now it's just not executed, end for customer is not really clear why

File: force-app/main/default/triggers/OpportunityTrigger.trigger [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        try{
+            if(Schema.sObjectType.Opportunity.isAccessible()){ //if user have access to object and fields
+                if(Trigger.isUpdate && Trigger.isAfter){
+                    OpportunityTriggerHelper.updatePlannedSalesWhenOpportunityClosedWon();
+                    OpportunityTriggerHelper.updateCarStatusWhenOpportunityStageChanged(Trigger.oldMap);
+                }else if(Trigger.isDelete && Trigger.isBefore){
+                    OpportunityTriggerHelper.deleteOpportunity();
+                }
+            }else{
+                system.debug('The user doesnt have access to this object');
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/OpportunityTrigger.trigger:16]: Use add error here instead

File: force-app/main/default/triggers/OpportunityTrigger.trigger [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    
+    Trigger_Control__c triggerSetting=Trigger_Control__c.getInstance('MakeActive');  
+    
+    if((triggerSetting!=null) && (triggerSetting.isActive__c)){
+        
+        try{
+            if(Schema.sObjectType.Opportunity.isAccessible()){ //if user have access to object and fields
+                if(Trigger.isUpdate && Trigger.isAfter){
+                    OpportunityTriggerHelper.updatePlannedSalesWhenOpportunityClosedWon();
+                    OpportunityTriggerHelper.updateCarStatusWhenOpportunityStageChanged(Trigger.oldMap);
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/OpportunityTrigger.trigger:11]: You are already using `Trigger.new` in your `updateCarStatusWhenOpportunityStageChanged` method, you can use Trigger.old/Trigger.oldMap in the same way, without passing it as a parameter. It make sense to do it like that if you will be calling these method from other places, where Trigger Context does not exists, but then it makes sense to pass Trigger.new as a parameter as well

File: force-app/main/default/triggers/PlannedSalesTrigger.trigger [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        try{
+            if(Schema.sObjectType.Planned_Sales__c.isAccessible()){//if user have access to object and fields
+                if(Trigger.isInsert && Trigger.isAfter){
+                    PlannedSalesTriggerHelper.createPlannedSales();
+                }
+                if(Trigger.isUpdate && Trigger.isBefore){ // if isAfter than will be System.FinalException: Record is read-only 
+                    PlannedSalesTriggerHelper.updatePlannedSales(Trigger.new,Trigger.oldMap);
+                }
+            }else{
+                system.debug('The user doesnt have access to this object');
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/PlannedSalesTrigger.trigger:16]: Use addError() here

File: force-app/main/default/triggers/PlannedSalesTrigger.trigger [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+            }
+            
+        }catch(Exception e){
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/PlannedSalesTrigger.trigger:19]: More error handling to the handler methods

File: force-app/main/default/triggers/OpportunityTrigger.trigger [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+                }else if(Trigger.isDelete && Trigger.isBefore){
+                    OpportunityTriggerHelper.deleteOpportunity();
+                }
+            }else{
+                system.debug('The user doesnt have access to this object');
+            }
+            
+        }catch(Exception e){
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/OpportunityTrigger.trigger:19]: More error handling to the handler methods

File: force-app/main/default/triggers/OpportunityLineItemTrigger.trigger [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+            }
+        }catch(Exception e){
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/OpportunityLineItemTrigger.trigger:16]: More error handling to the handler methods

File: force-app/main/default/triggers/CarTrigger.trigger [hunk 13]
```diff
@@ -0,0 +1,24 @@
+trigger CarTrigger on Product2 (before insert,after update) {
+    
+    Trigger_Control__c triggerSetting=Trigger_Control__c.getInstance('MakeActive');  
+    
+    if((triggerSetting!=null) && (triggerSetting.isActive__c)){
+        try{
+            if(Schema.sObjectType.Product2.isAccessible()){//if user have access to object and fields
+                if(Trigger.isAfter && Trigger.isUpdate){
+                    CarTriggerHelper.updateCar(Trigger.oldMap);
```
Comment (SvAndrushchak) [inline force-app/main/default/triggers/CarTrigger.trigger:9]: You are already using `Trigger.new` in your updateCar method, you can use Trigger.old/Trigger.oldMap in the same way, without passing it as a parameter. It make sense to do it like that if you will be calling these method from other places, where Trigger Context does not exists, but then it makes sense to pass Trigger.new as a parameter as well

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+        try{
+            
+            List<Opportunity> updatedOpportunities=[SELECT Id,CloseDate,StageName,Amount,(SELECT Id,Product2.Condition__c FROM OpportunityLineItems) FROM Opportunity 
+                                                    WHERE StageName='Closed Won' AND Id IN :Trigger.New];
+            
+            Map<Integer,Integer> monthAndYear=new Map<Integer,Integer>();
+            
+            Map<String,Planned_Sales__c> plannedSales=new Map<String,Planned_Sales__c>();
+            
+            Map<Id,Planned_Sales__c> plannedSalesToUpdate=new Map<Id,Planned_Sales__c>();
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:20]: you don't really need to have it as a Map. The only thing you do is save records to update them, but never access them by key. List also takes less space then Map. Please change to the List<>

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+            
+            if(!updatedOpportunities.isEmpty()){
+                for(Opportunity opportunity:updatedOpportunities){
+                    
+                    monthAndYear.put(opportunity.CloseDate.month(),opportunity.CloseDate.year());
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:25]: We've discussed this part. It's not a good idea to have a map in that case, because it will be erasing values for years with same month, for example, if you add 2024-10-10 & 2025-10-10 to the map, in the end you will only have {10 => 2025}, 2024 values will be just erased and replaced with 2025. Use Set<> for bot Month and year instead

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+                        for(Opportunity opportunity:updatedOpportunities){
+                            String key1=opportunity.CloseDate.year()+'_'+opportunity.CloseDate.month();
+                            
+                            for(OpportunityLineItem opp:opportunity.OpportunityLineItems){
+                                String key2=key1+'_'+opp.Product2.Condition__c;
+                                
+                                if(plannedSales.containsKey(key2)){
+                                    Planned_Sales__c plannedSale=plannedSales.get(key2);
+                                    
+                                    if(plannedSale!=null){
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:49]: In the 46th `if(plannedSales.containsKey(key2))` line you check if map contains the key you are using in line 47 to get the value `Planned_Sales__c plannedSale=plannedSales.get(key2)`. In that case you don't need the check for `plannedSale!=null`, because if key exists, it's value can never be null. map.put('key', null) will throw an error

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+                    if(!plannedSalesToUpdate.isEmpty()){
+                        update plannedSalesToUpdate.values();
+                    }
+                }
+                
+                
+            }
+            
+            
+        }catch (System.NullPointerException e){
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:71]: Here you only catch NullPointerException, what about other exception types?

For example QueryException, ListException etc?

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+            system.debug(e);
+            System.debug('Error message: ' + e.getMessage()); 
+            System.debug('Stack trace: ' + e.getStackTraceString());
+        }
+    }
+    
+    public static void deleteOpportunity(){
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:78]: The code for deleteOpportunity and updatePlannedSalesWhenOpportunityClosedWon is 95% similar and is just a duplicate. Try to make one same method instead and make it more dynamic in the places where the is a difference between update and delete

For example:

...
```
if(Trigger.isUpdate()){
plannedSale.Count_of_Cars_Sold__c += deletedOpportunity.OpportunityLineItems.size();
plannedSale.Total_Amount_Earned__c += deletedOpportunity.Amount;
} else if(Trigger.isDelete()){
plannedSale.Count_of_Cars_Sold__c-=deletedOpportunity.OpportunityLineItems.size();
plannedSale.Total_Amount_Earned__c-=deletedOpportunity.Amount;
}
```
...

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+                    if(!newOpportunity.OpportunityLineItems.isEmpty()){
+                        for(OpportunityLineItem oli:newOpportunity.OpportunityLineItems){ 
+                            oli.Product2.Car_Status__c='Reserved';
+                            carsToUpdate.add(oli.Product2);
+                        }
+                    }
+                    
+                }else if(newOpportunity.StageName!=oldOpportunity.StageName && newOpportunity.StageName=='Closed Won'){
+                    
+                    String key=newOpportunity.CloseDate.month()+'_'+newOpportunity.CloseDate.year();
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:169]: This variable is never used, remove it

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+                    
+                    for(OpportunityLineItem oli:newOpportunity.OpportunityLineItems){
+                        oli.Product2.Car_Status__c='Sold';
+                        oli.Product2.RecordTypeId='012WU000002S418YAC';
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:173]: Ids are different between salesforce orgs, so their hardcode is not allowed AT ALL. You can query RecordType using SOQL instead

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+                        carsToUpdate.add(oli.Product2);
+                        
+                    }
+                }else if(newOpportunity.StageName!=oldOpportunity.StageName && newOpportunity.StageName=='Closed Lost'){
+                    for(OpportunityLineItem oli:newOpportunity.OpportunityLineItems){
+                        oli.Product2.Car_Status__c='Available';
+                        carsToUpdate.add(oli.Product2);
+                    }
+                }else if(oldOpportunity.StageName=='Contract Sent' && newOpportunity.StageName!='Closed Won' &&
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:182]: You should put newOpportunity.StageName!=oldOpportunity.StageName as a first condition to check, as you make change only when stage name is changed. all other conditions should be places after that, so they are checked only if it makes sense to check them (stage is changed)

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+                        carsToUpdate.add(oli.Product2);
+                    }
+                }else if(oldOpportunity.StageName=='Contract Sent' && newOpportunity.StageName!='Closed Won' && 
+                         newOpportunity.StageName!='Closed Lost' && newOpportunity.StageName!=oldOpportunity.StageName){
+                             for(OpportunityLineItem oli:newOpportunity.OpportunityLineItems){
+                                 oli.Product2.Car_Status__c='Available';
+                                 carsToUpdate.add(oli.Product2);
+                             }
+                         }
+                if(oldOpportunity.StageName=='Closed Won' && newOpportunity.StageName!=oldOpportunity.StageName){
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:189]: Same here, check newOpportunity.StageName!=oldOpportunity.StageName before other conditions. In this case you also need to change car status to Available. 

I would separate it into two different methods. Car_Status__c can be changed to Available in the current one, and planned sales should be moved to a separate method (SOLID -> S -> Single Responsibility Principle)

File: force-app/main/default/classes/OpportunityTriggerHelperTest.cls [hunk 23]
```diff
@@ -0,0 +1,441 @@
...
+ * @author            : Nazar Voinarovych
+ * @group             : 
+ * @last modified on  : 10-07-2024
+ * @last modified by  : Nazar Voinarovych
+**/
+
+@isTest
+public with sharing class OpportunityTriggerHelperTest {
+    
+    private static Trigger_Control__c createTriggerControl(){ //custom setting trigger control
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelperTest.cls:12]: You should create a TestDataFactory class and store all methods, used to make data in test classes there. It will also allow you to use that methods in different test classes without creating duplicate code

File: force-app/main/default/classes/OpportunityTriggerHelperTest.cls [hunk 24]
```diff
@@ -0,0 +1,441 @@
...
+        OpportunityLineItem opportunityProduct=new OpportunityLineItem(OpportunityId=opportunityId,Product2Id=product2Id,
+                                                                       Quantity=1,UnitPrice=12500);
+        return opportunityProduct;
+    }
+    
+    //test positive
+    @isTest
+    private static void testPositiveOpportunityClose(){
+        
+        Trigger_Control__c triggerControl=createTriggerControl();
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelperTest.cls:73]: Create data in the @TestSetup method.

Methods defined with the @TestSetup annotation are used for creating common test records that are available for all test methods in the class.

If a test class contains a test setup method, the testing framework executes the test setup method first, before any test method in the class. Records that are created in a test setup method are available to all test methods in the test class and are rolled back at the end of test class execution. If a test method changes those records, such as record field updates or record deletions, those changes are rolled back after each test method finishes execution. The next executing test method gets access to the original unmodified state of those records.

File: force-app/main/default/classes/CarTriggerHelper.cls [hunk 25]
```diff
@@ -0,0 +1,37 @@
+public with sharing class CarTriggerHelper {
+    
+    public static void updateCar(Map<Id,Product2> oldCars){
+        
+        List<Product2> updatedCars=[SELECT Id,Car_Status__c,Condition__c,Name,reservedInOpportunity__c FROM Product2 WHERE Id IN:Trigger.new];
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/CarTriggerHelper.cls:5]: You already have all this fields in Trigger.new, so you don't have to query it again

File: force-app/main/default/classes/CarTriggerHelper.cls [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+        
+        Set<Id> findCarsId=new Set<Id>();
+        List<Opportunity> opportunityToUpdate=new List<Opportunity>();
+        
+        for(Product2 car:updatedCars){
+            Product2 oldCar=oldCars.get(car.Id);
+            
+            if(oldCar!=null){
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/CarTriggerHelper.cls:13]: You run this method on update only, it will always have oldCar record, so you can remove this if statement

File: force-app/main/default/classes/CarTriggerHelper.cls [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+            }
+        }
+        if(!findCarsId.isEmpty()){
+            List<OpportunityLineItem> oppLineItems=[SELECT Id,Product2Id,Product2.Condition__c,Product2.Car_Status__c,Product2.reservedInOpportunity__c,
+                                                           Opportunity.StageName,Opportunity.showWarningMessage__c 
+                                       FROM OpportunityLineItem WHERE Opportunity.showWarningMessage__c=true AND  Product2Id IN:findCarsId];
+            if(!oppLineItems.isEmpty()){
+            for(OpportunityLineItem oppLineItem:oppLineItems){
+                oppLineItem.Opportunity.showWarningMessage__c=false;
+                opportunityToUpdate.add(oppLineItem.Opportunity);
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/CarTriggerHelper.cls:26]: In this part you will add duplicates to opportunityToUpdate list if you have 2 or more line items for one opportunity. 

It will throw `System.ListException: Duplicate id in list` later on update.

In this case you wanna use a map to store updated opportunities in the way OpportunityId => Opportunity and then update map.values() list or save updated ids somewhere and then check if you already processed them

File: force-app/main/default/classes/CarTriggerHelperTest.cls [hunk 28]
```diff
@@ -0,0 +1,209 @@
+@isTest
+public with sharing class CarTriggerHelperTest {
+    
+    private static Trigger_Control__c createTriggerControl(){ //custom setting trigger control
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/CarTriggerHelperTest.cls:4]: Same here, move all data creation to TestDataFactory and call it on @TestSetup

File: force-app/main/default/classes/CarTriggerHelperTest.cls [hunk 29]
```diff
@@ -0,0 +1,209 @@
...
+    
+    private static OpportunityLineItem createOpportunityProduct(Id product2Id,Id opportunityId){
+        OpportunityLineItem opportunityProduct=new OpportunityLineItem(OpportunityId=opportunityId,Product2Id=product2Id,
+                                                                       Quantity=1,UnitPrice=12500);
+        return opportunityProduct;
+    }
+    
+    //test positive
+    @isTest
+    private static void testPositiveUpdateCar(){
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/CarTriggerHelperTest.cls:57]: You don't need a test for a single positive test if you already have a bulk one

File: force-app/main/default/classes/CarTriggerHelperTest.cls [hunk 30]
```diff
@@ -0,0 +1,209 @@
...
+        }
+        update opportunities;
+        
+        for(Product2 car:cars){
+            car.Car_Status__c='Available';
+        }
+        update cars;
+        
+        
+        Test.StopTest();
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/CarTriggerHelperTest.cls:206]: Don't see any asserts here. You have to check that all records are in the expected state after the test

File: force-app/main/default/classes/OpportunityLineItemTriggerHelper.cls [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+        
+        for(OpportunityLineItem oli:newOpportunityLineItems){
+            productsId.add(oli.Product2Id);
+        }
+        
+        if(!newOpportunityLineItems.isEmpty()){
+                for(OpportunityLineItem oli:newOpportunityLineItems){
+                    if((oli.Product2.Car_Status__c=='Reserved' || oli.Product2.Car_Status__c=='Sold')&& oli.Product2.reservedInOpportunity__c==true ){
+                        oli.Opportunity.showWarningMessage__c=true;
+                        opportunityToUpdate.add(oli.Opportunity);
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityLineItemTriggerHelper.cls:20]: Same as for CarTriggerHelper.updateCart method. It will throw an error on update, if list will have duplicates

File: force-app/main/default/classes/OpportunityLineItemTriggerHelper.cls [hunk 32]
```diff
@@ -0,0 +1,60 @@
...
+            for(OpportunityLineItem oli:deletedOpportunityLineItems){
+                productIdOfDeleteOppLineItems.add(oli.Product2Id);
+            }   
+                if(!deletedOpportunityLineItems.isEmpty()){
+                    for(OpportunityLineItem oli:deletedOpportunityLineItems){
+                        system.debug(oli);
+                        if((oli.Product2.Car_Status__c=='Sold' || oli.Product2.Car_Status__c=='Reserved') 
+                           && (oli.Opportunity.showWarningMessage__c==true) && oli.Product2.reservedInOpportunity__c==true){
+                            oli.Opportunity.showWarningMessage__c =false;
+                            opportunityToUpdate.add(oli.Opportunity);
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityLineItemTriggerHelper.cls:48]: Same as for CarTriggerHelper.updateCart method. It will throw an error on update, if list will have duplicates

File: force-app/main/default/classes/PlannedSalesTriggerHelper.cls [hunk 33]
```diff
@@ -0,0 +1,138 @@
...
+ * @group             : 
+ * @last modified on  : 10-07-2024
+ * @last modified by  : Nazar Voinarovych
+**/
+
+public with sharing class PlannedSalesTriggerHelper {
+    
+    public static void createPlannedSales(){
+        
+        List<Planned_Sales__c> createdPlannedSales=[SELECT Id,Year__c,Month__c,Car_Type__c,Total_Amount_Earned__c,Count_of_Cars_Sold__c
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/PlannedSalesTriggerHelper.cls:13]: You already have all of that in the Trigger.new. Don't need to query it again

File: force-app/main/default/classes/PlannedSalesTriggerHelper.cls [hunk 34]
```diff
@@ -0,0 +1,138 @@
...
+        
+        //Map<Integer,Integer> monthAndYear=new Map<Integer,Integer>();
+        Set<Integer> month=new Set<Integer>();
+        Set<Integer> year=new Set<Integer>();
+        Map<String,Planned_Sales__c> plannedSales=new Map<String,Planned_Sales__c>();
+        
+        
+        Map<Id,Planned_Sales__c> plannedSalesToUpdate=new Map<Id,Planned_Sales__c>();
+        
+        if(!newPlannedSales.isEmpty()){
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/PlannedSalesTriggerHelper.cls:82]: Trigger.new if never empty on update

File: force-app/main/default/classes/PlannedSalesTriggerHelper.cls [hunk 35]
```diff
@@ -0,0 +1,138 @@
...
+                } 
+            }
+ 
+        }
+        if(!plannedSalesToUpdate.isEmpty()){
+            update plannedSalesToUpdate.values();
+        }
+
+    }
+    public static void updatePlannedSales(List<Planned_Sales__c> newPlannedSales,Map<Id,Planned_Sales__c> oldPlannedSalesMap){
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/PlannedSalesTriggerHelper.cls:72]: This method literally do the same an the one on insert, just combine them instead of duplicating the code

File: force-app/main/default/classes/PlannedSalesTriggerHelper.cls [hunk 36]
```diff
@@ -0,0 +1,138 @@
...
+                        
+                        String key2=key+'_'+oli.Product2.Condition__c;
+                        
+                        if(relatedPlannedSales.containsKey(key2)){
+                            
+                            Planned_Sales__c plannedSale=relatedPlannedSales.get(key2);
+                            
+                            if(plannedSale!=null){
+                                plannedSale.Total_Amount_Earned__c+=opportunity.Amount;
+                                plannedSale.Count_of_Cars_Sold__c+=opportunity.OpportunityLineItems.size();
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/PlannedSalesTriggerHelper.cls:57]: Adding Opportunity line item amount on each LineItem iteration will not calculate if correctly

If you have 10 cars each 100k -> Opp.Amount  = 1 mil, 10 iterations will add 10 mil instead of 1 min

Another thing with Count_of_Cars_Sold__c. You have to only add cars with same Condition as Car Type on planned sales

File: force-app/main/default/classes/PlannedSalesTriggerHelper.cls [hunk 37]
```diff
@@ -0,0 +1,138 @@
...
+                 String key=opportunity.CloseDate.month()+'_'+opportunity.CloseDate.year();                           
+                 for(OpportunityLineItem oli:opportunity.OpportunityLineItems){
+                    String key2=key+'_'+oli.Product2.Condition__c;
+                     
+                     
+                     if(plannedSales.containsKey(key2)){
+                         
+                         Planned_Sales__c plannedSale=plannedSales.get(key2);
+                         if(plannedSale!=null){
+                             plannedSale.Total_Amount_Earned__c+=opportunity.Amount;
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/PlannedSalesTriggerHelper.cls:116]: Adding Opportunity line item amount on each LineItem iteration will not calculate if correctly

If you have 10 cars each 100k -> Opp.Amount = 1 mil, 10 iterations will add 10 mil instead of 1 min

Another thing with Count_of_Cars_Sold__c. You have to only add cars with same Condition as Car Type on planned sales

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 38]
```diff
@@ -0,0 +1,217 @@
...
+                            String key1=opportunity.CloseDate.year()+'_'+opportunity.CloseDate.month();
+                            
+                            for(OpportunityLineItem opp:opportunity.OpportunityLineItems){
+                                String key2=key1+'_'+opp.Product2.Condition__c;
+                                
+                                if(plannedSales.containsKey(key2)){
+                                    Planned_Sales__c plannedSale=plannedSales.get(key2);
+                                    
+                                    if(plannedSale!=null){
+                                        plannedSale.Total_Amount_Earned__c+=opportunity.Amount;
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:50]: Adding Opportunity line item amount on each LineItem iteration will not calculate if correctly

If you have 10 cars each 100k -> Opp.Amount = 1 mil, 10 iterations will add 10 mil instead of 1 min

Another thing with Count_of_Cars_Sold__c. You have to only add cars with same Condition as Car Type on planned sales

File: force-app/main/default/classes/OpportunityTriggerHelper.cls [hunk 39]
```diff
@@ -0,0 +1,217 @@
...
+                
+                if(!relatedPlannedSales.isEmpty()){
+                    for(Opportunity deletedOpportunity:deletedOpportunities){
+                        String key=deletedOpportunity.CloseDate.month()+'_'+deletedOpportunity.CloseDate.year();
+                        for(OpportunityLineItem oli:deletedOpportunity.OpportunityLineItems){
+                            String key2=key+'_'+oli.Product2.Condition__c;
+                            if(relatedPlannedSales.containsKey(key2)){
+                                Planned_Sales__c plannedSale=relatedPlannedSales.get(key2);
+                                if(plannedSale!=null){
+                                    plannedSale.Count_of_Cars_Sold__c-=deletedOpportunity.OpportunityLineItems.size();
```
Comment (SvAndrushchak) [inline force-app/main/default/classes/OpportunityTriggerHelper.cls:112]: Adding Opportunity line item amount on each LineItem iteration will not calculate if correctly

If you have 10 cars each 100k -> Opp.Amount = 1 mil, 10 iterations will add 10 mil instead of 1 min

Another thing with Count_of_Cars_Sold__c. You have to only add cars with same Condition as Car Type on planned sales

File: force-app/main/default/classes/CarTriggerHelper.cls (hunk 25 above)
Comment (SvAndrushchak) [reply force-app/main/default/classes/CarTriggerHelper.cls:5]: List<Product2> updatedCars = (List<Product2>)Trigger.new;
